### DISCUSSION TITLE
Can the Grid handle large amounts of records?

### DISCUSSION THREAD

**Discussion by donnyv** [AvaloniaUI/Avalonia#5824] [Q&A]
I'm new to Avalonia and I'm researching to see if Avalonia can be used with a new project I'm working on. Can the Grid handle large amounts of records like 100k+? Also does it use virtual scrolling? Can it handle infinite scrolling?

**Comment by sn4k3** [comment]
For me loading 23k records are slow +/- 20s, i'm using the standard call

**Reply by maxkatz6** [reply]
In your case most likely something wrong with virtualization or you are using complex template columns.

If I open ControlCatalog and modify this line https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/DataGridPage.xaml.cs#L20
to have more rows (~1000*200=200k):
```csharp
var collectionView1 = new DataGridCollectionView(Enumerable.Range(0, 1000).SelectMany(_ => Countries.All).ToArray());
```
It works just fine with 1-2 seconds to sort all the data. Scrolling is fast.
I don't know if it supports infinite scrolling though.

**Reply by sn4k3** [reply]
I'm computing the items first.
In the end i do a ObservableCollection.AddRange(list), and that method is taking 23s (with debugger on)
The class is: https://github.com/sn4k3/UVtools/blob/6cfedea4cc36200caab87301e0dd121e97ddc8f2/UVtools.Core/Layer/LayerIssue.cs#L242
My datagrid: https://github.com/sn4k3/UVtools/blob/master/UVtools.WPF/MainWindow.axaml#L795

![image](https://user-images.githubusercontent.com/113281/115635505-c4633f00-a303-11eb-9b91-8df810f73937.png)

**Reply by maxkatz6** [reply]
Well, it's unrelated to datagrid. In your case instead of iterating over array, you have IEnumerator instead. 
And I don't remember AddRange method in standard ObservableCollection, so most likely you are using some extension method that just invokes Add per item, which executes INotifyCollectionChanged handlers. Per item. So yeah, not a surprise that it works slow.

**Reply by sn4k3** [reply]
Yes it's an extension method, doing exactly that...
Is there a method to insert a range without trigger a notify per item?

**Reply by maxkatz6** [reply]
Something like this
https://stackoverflow.com/questions/670577/observablecollection-doesnt-support-addrange-method-so-i-get-notified-for-each

It's long living BCL issue with ObservableCollection.

**Reply by sn4k3** [reply]
I tried that alternative but Datagrid only show 1 record after the AddRange. But it now takes 7ms
![image](https://user-images.githubusercontent.com/113281/115637549-4d7c7500-a308-11eb-91ee-66e113a2c205.png)

**Reply by maxkatz6** [reply]
Have you tried it with Reset operation?
```
        OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
```

Ref: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs#L3403

**Reply by sn4k3** [reply]
It worked! now 99ms for the 23k under the debugger. Thank you!

**Reply by donnyv** [reply]
This went a little tangent. But happy it fixed someone's problem and answered my question. Looks like it will handle large amounts of records fine 

Thanks!
